### PR TITLE
Refactor `Replicate#getLastRelevantStatus` to remove the `Optional` return type

### PR DESCRIPTION
**Pull request by mcornaton**
PR goal: fix [SonarCloud detected bugs on `Replicate#getLastRelevantStatus`](https://sonarcloud.io/project/issues?resolved=false&severities=MAJOR&sinceLeakPeriod=true&types=BUG&id=com.iexec.core%3Aiexec-core)

### CODE REVIEW

File: src/main/java/com/iexec/core/replicate/Replicate.java [hunk 1]
```diff
@@ -64,7 +63,7 @@ public ReplicateStatus getCurrentStatus() {
     }
 
     @JsonIgnore
-    public Optional<ReplicateStatus> getLastRelevantStatus() {  // FIXME: remove Optional and add a no-args constructor
+    public ReplicateStatus getLastRelevantStatus() {
         // ignore cases like: WORKER_LOST and RECOVERING
 
         List<ReplicateStatus> statusList = getStatusUpdateList().stream()
```
Comment (jbern0rd) [inline src/main/java/com/iexec/core/replicate/Replicate.java:69]: Not in the PR originally, but could we globally replace `getStatusUpdateList()` with direct access to `statusUpdateList` field ?

File: src/main/java/com/iexec/core/replicate/Replicate.java [hunk 2]
```diff
@@ -19,28 +19,30 @@
...
+import static com.iexec.common.replicate.ReplicateStatusUpdate.poolManagerRequest;
 
 
 @Data
 @NoArgsConstructor
 @JsonIgnoreProperties(ignoreUnknown = true)
 public class Replicate {
 
-    private List<ReplicateStatusUpdate> statusUpdateList;
+    private List<ReplicateStatusUpdate> statusUpdateList = new ArrayList<>(
```
Comment (jbern0rd) [inline src/main/java/com/iexec/core/replicate/Replicate.java:42]: Add `FIXME` to pass this field `final` for immutability
To be done in another PR, breaks many tests in detector tests, `TaskUpdateManagerTest`, `ReplicateTests` and `ReplicateSupplyServiceTests`.
More analysis is required to replace calls to `setStatusUpdateList()`.

File: src/main/java/com/iexec/core/replicate/Replicate.java [hunk 3]
```diff
@@ -77,11 +76,11 @@ public Optional<ReplicateStatus> getLastRelevantStatus() {  // FIXME: remove Opt
 
         for (int i = statusList.size() - 1; i >= 0; i--) {
             if (!ignoredStatuses.contains(statusList.get(i))) {
-                return Optional.of(statusList.get(i));
+                return statusList.get(i);
             }
         }
 
-        return Optional.empty();
+        throw new NoReplicateStatusException(chainTaskId);
```
[the remark this answers is not in the document]
Comment (jbern0rd) [reply src/main/java/com/iexec/core/replicate/Replicate.java:83]: To improve on this, the mentioned `FIXME` must be implemented to have immutable `statusUpdateList`.
To be done in a future PR.

File: src/main/java/com/iexec/core/replicate/Replicate.java (hunk 1 above)
Comment (mcornaton) [reply src/main/java/com/iexec/core/replicate/Replicate.java]: Sure! Please see https://github.com/iExecBlockchainComputing/iexec-core/pull/524/commits/dbcab6ac8f5b20da31a08674bd523f3c65bf1409.
(Issue with the commit name but that's the right one! :) )

File: src/main/java/com/iexec/core/replicate/Replicate.java (hunk 2 above)
Comment (mcornaton) [reply src/main/java/com/iexec/core/replicate/Replicate.java:43]: Added there! https://github.com/iExecBlockchainComputing/iexec-core/pull/524/commits/dbcab6ac8f5b20da31a08674bd523f3c65bf1409

File: src/main/java/com/iexec/core/contribution/ContributionHelper.java [hunk 4]
```diff
@@ -92,12 +85,8 @@ static Set<String> getDistinctContributions(List<Replicate> replicates) {
...
         for (Replicate replicate : replicates) {
 
-            Optional<ReplicateStatus> lastRelevantStatus = replicate.getLastRelevantStatus();
-            if (lastRelevantStatus.isEmpty()) {
-                continue;
-            }
-
-            if (lastRelevantStatus.get().equals(ReplicateStatus.CONTRIBUTED)) {
+            ReplicateStatus lastRelevantStatus = replicate.getLastRelevantStatus();
+            if (lastRelevantStatus.equals(ReplicateStatus.CONTRIBUTED)) {
```
[the remark this answers is not in the document]
Comment (mcornaton) [reply src/main/java/com/iexec/core/contribution/ContributionHelper.java]: Using `==` is a bit better for enums. Please see https://github.com/iExecBlockchainComputing/iexec-core/pull/524/commits/566e5328789fb697250547678e5f34951279c135 :)

File: src/main/java/com/iexec/core/replicate/ReplicateSupplyService.java [hunk 5]
```diff
@@ -408,12 +400,8 @@ private Optional<TaskNotificationType> recoverReplicateInRevealPhase(Task task,
...
         String walletAddress = replicate.getWalletAddress();
 
-        if (replicate.getLastRelevantStatus().isEmpty()) {
-            return Optional.empty();
-        }
-
-        boolean isInStatusContributed = replicate.getLastRelevantStatus().get().equals(ReplicateStatus.CONTRIBUTED);
-        boolean didReplicateStartRevealing = replicate.getLastRelevantStatus().get().equals(ReplicateStatus.REVEALING);
+        boolean isInStatusContributed = replicate.getLastRelevantStatus().equals(ReplicateStatus.CONTRIBUTED);
+        boolean didReplicateStartRevealing = replicate.getLastRelevantStatus().equals(ReplicateStatus.REVEALING);
```
Comment (jbern0rd) [inline src/main/java/com/iexec/core/replicate/ReplicateSupplyService.java:404]: After our discussion, maybe ?
```suggestion
        boolean isInStatusContributed = replicate.getLastRelevantStatus() == ReplicateStatus.CONTRIBUTED;
        boolean didReplicateStartRevealing = replicate.getLastRelevantStatus() == ReplicateStatus.REVEALING;
```

Comment (mcornaton) [reply src/main/java/com/iexec/core/replicate/ReplicateSupplyService.java]: Should now be clean everywhere :)
https://github.com/iExecBlockchainComputing/iexec-core/pull/524/commits/7e35e70e9bf0d32693a888ee7d39e5e5596d8ed0

**Review by jbern0rd — APPROVED** [review]
Nice work, many thanks
